client/src/main/java/org/apache/nemo/client/beam/NemoPipelineResult.java

```diff
@@ -49,10 +64,19 @@ public State cancel() throws IOException {
 
   @Override
   public State waitUntilFinish(final Duration duration) {
-    throw new UnsupportedOperationException();
-    // TODO #208: NemoPipelineResult#waitUntilFinish hangs
-    // Previous code that hangs the job:
-    // return (State) super.waitUntilJobFinish(duration.getMillis(), TimeUnit.MILLISECONDS);
+    final boolean finished;
+    try {
+      finished = this.jobDone.await(duration.getMillis(), TimeUnit.MILLISECONDS);
+    } catch (final InterruptedException e) {
+      throw new RuntimeException(e);
+    }
+    if (finished) {
+      LOG.info("Job successfully finished before timeout of {}ms, while waiting until finish", duration);
+    } else {
+      LOG.info("Job timed out before {}ms, while waiting until finish.", duration);
+      JobLauncher.shutdown();
```

Comment:
I think the shutdown should be called in `PipelineResult.cancel()`. Users can cancel if it is not finished after the duration. This method should only return the state. (e.g., job still running or finished?)